### PR TITLE
store/tikv: add a context in BackOffer to support cancel operation

### PR DESCRIPTION
**Pull request by tiancaiamao**
1. add a `context.Context` in `BackOffer` struct
2. `WithCancel` provides a way to release resources as soon as possible.

in the `doBatch` function, stop sending other requests after receiving first error. 
ref: https://github.com/pingcap/tidb/issues/1847

----------

this commit introduce a context.Context but not thoroughly,
later we should add more support for context.Context, use less `context.Background()` and add more

```
select {
    case <-ctx.Done():
}
```

### CODE REVIEW

File: store/tikv/txn_committer.go [hunk 1]
```diff
@@ -159,6 +161,7 @@ func (c *txnCommitter) doBatches(bo *Backoffer, batches []batchKeys, f func(*Bac
 	for i := 0; i < len(batches); i++ {
 		if e := <-ch; e != nil {
 			log.Warnf("txnCommitter doBatches failed: %v, tid: %d", e, c.startTS)
+			cancel()
```
Comment (tiancaiamao) [inline store/tikv/txn_committer.go]: the core of changes is just this line

**Comment by ngaut** [comment]
PTAL @disksing

File: store/tikv/backoff.go [hunk 2]
```diff
@@ -128,15 +129,24 @@ type Backoffer struct {
...
 	return &Backoffer{
 		maxSleep: maxSleep,
+		ctx:      ctx,
 	}
 }
 
+// WithCancel returns a cancel function which, when called, would cancel backoffer's context.
+func (b *Backoffer) WithCancel() context.CancelFunc {
+	var cancel context.CancelFunc
+	b.ctx, cancel = context.WithCancel(b.ctx)
```
Comment (siddontang) [inline store/tikv/backoff.go]: why WithCancel overwrites parent b.ctx? 
We should return the child ctx and cancel function for outer usage.

Comment (tiancaiamao) [inline store/tikv/backoff.go]: standard use context.Context follow this pattern:

```
func f(ctx context.Context, ...) {
    ctx1, cancel := context.WithCancel(ctx)
    g(ctx1, ...)
}
```
our functions all use `bo *BackOffer` as the first argument, so I put ctx in BackOffer struct.

WithCancel won't overwrite parent context, it just wrap the parent context...

```
func f(bo *Backoffer, ...) {
  cancel := bo.WithCancel(ctx)
  g(bo, ...)
}
```

this code works just the same.

**Comment by tiancaiamao** [comment]
As closure is not comparable in Go, I can't write:

```
if f == c.prewriteSingleRegion {
    cancel()
}
```

So I add a flag cancelOnFirstError in Backoffer, set that flag to true when `prewriteKeys`.
if flag cancelOnFirstError is not set, WithCancel return a nop operation.

@disksing

**Comment by disksing** [comment]
I think we can add a `cancelOnFirstError` flag parameter to `doBatches`, and only enable it for prewrite.

**Comment by disksing** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @ngaut @shenli @coocood

File: store/tikv/txn_committer.go [hunk 3]
```diff
@@ -235,6 +244,13 @@ func (c *txnCommitter) prewriteSingleRegion(bo *Backoffer, batch batchKeys) erro
 				return errors.Trace(err)
 			}
 		}
+
+		select {
```
Comment (coocood) [inline store/tikv/txn_committer.go]: This is redundant, cancel is already checked in `SendKVReq`.

**Comment by coocood** [comment]
LGTM

**Pull request merged**